### PR TITLE
Add RDMA write support to EFA provider

### PR DESCRIPTION
**Pull request by YonatanNachum**
EFA devices are now support RDMA write, make minor refactoring of EFA send inner function for creating wqe, set remote address and set immediate, and add RDMA write support.

### CODE REVIEW

**Comment by gal-pressman** [comment]
Where is the kernel side?

**Comment by YonatanNachum** [comment]
Kernel side is dependent on data polling patch, will be glad to get reviews for both:
[PR](https://github.com/linux-rdma/rdma-core/pull/1312)
[kernel patch](https://patchwork.kernel.org/project/linux-rdma/patch/20230219081328.10419-1-mrgolin@amazon.com/)

thanks

**Comment by jgunthorpe** [comment]
This still seems like the kernel side hasn't been posted, I don't see it on patchworks.

**Comment by mrgolin** [comment]
We've found another needed change, will sent kernel patch soon.

**Comment by YonatanNachum** [comment]
[kernel patch](https://lore.kernel.org/linux-rdma/20230404154313.35194-1-ynachum@amazon.com/)

Thanks.

**Comment by jgunthorpe** [comment]
This needs to be rebased

**Comment by YonatanNachum** [comment]
@jgunthorpe 
Rebased branch and updated kernel headers.

Thanks.

File: providers/efa/man/efadv_query_device.3.md [hunk 1]
```diff
@@ -76,6 +76,9 @@ struct efadv_device_attr {
 		Reading source address (SGID) from receive completion descriptors is supported.
 		Valid only for unknown AH.
 
+	EFADV_DEVICE_ATTR_CAPS_RDMA_WRITE:
+		RDMA write is supported
+
 *max_rdma_size*
```
Comment (gal-pressman) [inline providers/efa/man/efadv_query_device.3.md:82]: Does this apply to both RDMA read and write now?

Comment (YonatanNachum) [reply providers/efa/man/efadv_query_device.3.md:82]: Hey,
No this does not apply for RDMA read. For RDMA read you must get also EFADV_DEVICE_ATTR_CAPS_RDMA_READ to know it is supported.

Comment (gal-pressman) [reply providers/efa/man/efadv_query_device.3.md:82]: I'm referring to max_rdma_size.

Comment (YonatanNachum) [reply providers/efa/man/efadv_query_device.3.md:82]: Hey, sorry.
Yes, the max RDMA size is identical for read and write.

**Comment by YonatanNachum** [comment]
@jgunthorpe
Hey, kind reminder for this patch.

Thank you.

**Pull request merged**